feat: add vim.filetype.get_option()

lewis6991 opened the pull request:
### Problem

It is not possible to get the default option value for a specific filetype. This is required for a number of applications, including working with treesitter injections, e.g: automatically updating `'commentstring'` based on cursor position.

### Solution

- Add `filetype` option to `nvim_get_option_value()`  to get the default option value for a specific filetype.
- Add `vim.filetype.get_option(filetype, option)` as a wrapper around `nvim_get_option_value()` with caching.

---

<details>
<summary>Use this to set a context aware commentstring</summary>

```lua

local function get_lang()
  local ok, parser = pcall(vim.treesitter.get_parser)
  if not ok then
    return
  end

  local cpos = api.nvim_win_get_cursor(0)
  local row, col = cpos[1] - 1, cpos[2]
  local range = { row, col, row, col + 1 }

  local lang ---@type string?
  parser:for_each_child(function(tree, lang_)
    if tree:contains(range) then
      lang = lang_
      return
    end
  end)

  return lang
end

local function enable_commenstrings()
  -- Any better events?
  api.nvim_create_autocmd({'CursorMoved', 'CursorMovedI'}, {
    buffer = 0,
    callback = function()
      local lang = get_lang() or vim.bo.filetype
      if lang == 'comment' then
        lang = vim.bo.filetype   -- Exclude this
      end

      local commentstring = vim.filetype.get_option(lang, 'commentstring')

      if commentstring ~= vim.bo.commentstring then
        vim.bo.commentstring = commentstring
      end
    end
  })
end

api.nvim_create_autocmd('FileType', {
  callback = function()
    if not pcall(vim.treesitter.start) then
      return
    end

    enable_commenstrings()
  end
})

```

</details>

lewis6991 committed:
feat: add `vim.filetype.get_option()`

clason commented:
@echasnovski @numToStr

echasnovski commented:
Nice! Implementation seems a bit hacky, but can't suggest any better. Will definitely use it.

clason commented:
Are there other filetype-local options that could be interesting (and thus warrant a generic vim.treesitter.get_lang_option)?

lewis6991 commented:
It is what it is. Fundementally we need to load `runtime/ftplugin/<filetype>.vim` which needs a buffer to pollute information into.

lewis6991 commented:
> Are there other filetype-local options that could be interesting (and thus warrant a generic vim.treesitter.get_lang_option)?

Good question. Maybe.

lewis6991 commented:
Quick look:

```vim
setlocal cinkeys-=0#
setlocal indentkeys-=0#
setlocal include=^\\s*\\(from\\\|import\\)
setlocal define=^\\s*\\(def\\\|class\\)
setlocal includeexpr=...
setlocal suffixesadd=.py
setlocal comments=b:#,fb:-
setlocal commentstring=#\ %s
setlocal omnifunc=python3complete#Complete
setlocal keywordprg=python3\ -m\ pydoc
setlocal isk+=#
```

echasnovski commented:
> Are there other filetype-local options that could be interesting (and thus warrant a generic vim.treesitter.get_lang_option)?

'comments' can be an important one for commenting plugins.

echasnovski commented:
> It is what it is. Fundementally we need to load `runtime/ftplugin/<filetype>.vim` which needs a buffer to pollute information into.

One downside to this I've realized is that it can heavily increase value of new buffer numbers. Not a big deal but might be not obvious for a general user which navigates using buffer numbers.

lewis6991 commented:
Either that or we cache the commentstring values.

Or even more radical, do something similar to `filetype.lua` and just maintain a big table.

clason commented:
> Quick look:
> 
> ```viml
> setlocal cinkeys-=0#
> setlocal indentkeys-=0#
> setlocal include=^\\s*\\(from\\\|import\\)
> setlocal define=^\\s*\\(def\\\|class\\)
> setlocal includeexpr=...
> setlocal suffixesadd=.py
> setlocal comments=b:#,fb:-
> setlocal omnifunc=python3complete#Complete
> setlocal keywordprg=python3\ -m\ pydoc
> setlocal isk+=#
> ```
> 
> I don't think we heavily use many (any?) of these.

Omnifunc sounds very useful but would need to be handled at a lower level; same with idk and keywordprg.

Would have to be something that plugins need to know about. (If we had comment in core...)

clason commented:
> Either that or we cache the commentstring values.
> 
> Or even more radical, do something similar to `filetype.lua` and just maintain a big table.

_That_ should definitely be more generic then... (But hard to make backward compatible.)

If we go that route, we should integrate _all_ such options into filetype.lua, which already supports setting filetype options (because filetype.vim does, because why not add another mechanism on top of existing ones).

clason commented:
> Either that or we cache the commentstring values.

I think that would be a reasonable and natural optimization -- not necessarily for this PR, and not until someone complains about performance...

lewis6991 commented:
Caching is easy enough and as `commentstring` isn't dynamic for a filetype in any way, we can keep the cache live for the whole session and not worry about invalidating. I think I'd rather do that than have a buffer lying around.

lewis6991 commented:
> If we go that route, we should integrate _all_ such options into filetype.lua, which already supports setting filetype options (because filetype.vim does, because why not add another mechanism on top of existing ones).

Well...as per my comments on the other PR, I'm not sure `define` or `include` should even exist, so maybe not **all**.

seblj commented on runtime/lua/vim/treesitter.lua:
```diff
@@ -664,4 +664,35 @@ function M.foldexpr(lnum)
+  parser:for_each_child(function(tree, lang)
+    if tree:contains(range) then
+      local buf = a.nvim_create_buf(false, true)
+      vim.bo[buf].filetype = lang
```
Won't this potentially be an issue because of the use of `lang`? For example in markdown, the lang may be `markdown_inline`, and it will then default to commentstring `/*%s*/`

lewis6991 replied:
But markdown_inline will have an injection so we'd use that.

lewis6991 replied:
Just define a commentstring for markdown_inline then. It hasn't a big problem.

justinmk commented:
> I think I'd rather do that than have a buffer lying around.

fwiw, it can be `:bdelete` 'd and revived whenever needed (set 'bufhidden=hide' for good measure). 

The interface matters more, and it sounds like we are heading in a good direction here. `vim.treesitter.get_option()` seems actually more _intuitive_ than having to look around to see which options are and aren't "provided" by `vim.treesitter`. For that reason I would even suggest _hiding_ `vim.treesitter.foldexpr` and encapsulating it in `vim.treesitter.get_option()` , to further emphasize the pattern--though foldexpr may be special-cased internally?

mg979 commented on runtime/lua/vim/treesitter.lua:
```diff
@@ -664,4 +664,35 @@ function M.foldexpr(lnum)
+  local commentstring
+  parser:for_each_child(function(tree, lang)
+    if tree:contains(range) then
+      local buf = a.nvim_create_buf(false, true)
```
I suggest using a script-local `tempbuf` variable, then doing:
```lua
tempbuf = tempbuf and a.nvim_buf_is_valid(tempbuf) and tempbuf or a.nvim_create_buf(false, true)
```

without deleting that buffer afterwards, so that it can be reused, but it's created again if it gets deleted for some reason (like user doing some radical cleanup operation on all buffers).

lewis6991 commented:
`foldexpr` is a window-local option so it doesn't quite work here. Only buffer-local options make sense.

clason commented:
Thoughts on the generic function:
1. `vim.treesitter.get_option()` is not great since it can be interpreted to mean "get _treesitter_ option" (whatever those are or will be). Maybe `get_local_option` or `get_buf_option`?
2. The idea is to have a generic interface that will _not_ require hard-coding some "blessed" options, so whether, e.g., `include` is pointless or not makes no difference. This is assuming this can be implemented transparently (just returning `vim.bo[buf][bufoption]`, where `bufoption` is an argument to `get_local_option` -- with or without validation), otherwise it becomes a harder sell.

echasnovski commented:
> 2\. The idea is to have a generic interface that will _not_ require hard-coding some "blessed" options, so whether, e.g., `include` is pointless or not should make no difference. This is assuming this can be implemented transparently (just returning `vim.bo[buf][bufoption]`, where `bufoption` is an argument to `get_local_option` -- with or without validation), otherwise it becomes a harder sell.

I think my initial expectation for allowed options in a function `vim.treesitter.get_buf_option()` would be the ones for which `vim.api.nvim_get_option_info(option_name).scope` is `'buf'`.

clason commented:
Sure, that would be one method of validation.

lewis6991 commented:
These aren't just buffer-local options as these injections aren't real buffers. This is just how vim provides some of these filetype specific options and is strictly finite since I don't expect we'll be adding more.

Going forward, I expect filetype specific properties to be managed in a different system separate from the concept of buffers (like `lang`). E.g. I don't imagine we'll be adding `contextexpr`, or should we?

clason commented:
Thinking about it, to pull this off properly we also need a reverse mapping "parser name -> filetype" to look up the correct commentstring for languages where the parser name is different from the filetype (`latex` vs. `tex`; `bibtex` vs. `bib`; `markdown_inline` vs. `markdown`, ...)

(Or any other way of associating language to canonical filetype, no matter how we end up implementing the `filetype -> parser name` mapping.)

clason commented:
> This is just how vim provides some of these filetype specific options and is strictly finite since I don't expect we'll be adding more.

Yes, but users could, and would expect them to be resolved? I just think it's simpler to document (and probably implement) that the function would "return any buffer-local option as if the region was a freshly opened buffer of that filetype".

lewis6991 commented:
Been thinking about this, and there may be some performance concerns. This is because we need to fire the `FileType` autocmd which is dynamic and can source any number of files. For infrequent calls, like to fetch the commentstring when commenting a single line, this wouldn't be a problem. But for example, if you want to comment every line in a file correctly, a naive implementation (on both sides) would result in `FileType` being fired N times and source O(n) files.

A good implementation of a commenting function could know about this perf characteristic and decide to keep a short lived local cache, however I would expect other implementation/applications to not take this into account.

Due to how dynamic autocmds make this, I don't think it's possible to implement a private cache in `get_option`, or atleast not easily as we'd need to keep track of sourced file paths (difficult) and autocmd entries for `FileType` (easy).

The only think I can think of would be to just document that this new `get_option` function is expensive and shouldn't be called often.

clason commented:
I agree with the analysis, but since we have to add a caveat either way, we could also say "value is cached and may not update in every situation"? (And wait for complaints to see if these situations are worth supporting.)

lewis6991 commented:
I've gone ahead and just implemented a cache. It's strictly only looks in `ftplugin/*` so anything outside of that is out of scope.

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -2203,6 +2203,18 @@ add({filetypes})                                          *vim.filetype.add()*
                      example).
 
+get_option({filetype}, {option})                   *vim.filetype.get_option()*
+    Get the default option value for a filetype set bt ftplugin Example: >lua
```
putting this on vim.filetype is a good idea!

```suggestion
    Get the default option value for a filetype set by ftplugin Example: >lua
```

clason commented on runtime/lua/vim/treesitter.lua:
```diff
@@ -658,4 +658,36 @@ function M.foldexpr(lnum)
+---             - pos table|nil 0-indexed (row, col) tuple. Defaults to cursor position in the
+---                             current window. Required if {bufnr} is not the current buffer
+---@return string|nil
+function M.get_lang(opts)
```
This function already exists elsewhere:
https://github.com/neovim/neovim/blob/d1e0f7454b5fc61d26db5af5ce00c1894e7c49fc/runtime/lua/vim/treesitter/language.lua#L8-L12

(Guess who added that last month ;))

clason replied:
(Can probably be combined into one interface by adding `filetype` to the `opts` keys here and marking the other function as private via `_get_lang_from_ft`.)

lewis6991 replied:
`get_lang()` gets removed in #22643. Instead, we will use `vim.filetype.get_option(filetype, 'parser')` as a direct replacement.

lewis6991 replied:
FYI, this is why I was trying to drill down if `ftplugin` was the way to go to use as a namespace.

clason replied:
> get_lang() gets removed in https://github.com/neovim/neovim/pull/22643. Instead, we will use vim.filetype.get_option(filetype, 'parser') as a direct replacement.

So that PR is a prerequisite for this one? (Asking because that duplication is what's failing all the builds.)

lewis6991 replied:
Also, I'm planning on changing the nomenclature to be "parser" instead of "lang". TS uses "language" but I think that's too confusing with out "filetype". I think it's much easier to understand the concept of a "parser" being able to parser multiple "filetypes".

clason replied:
As long as people coming from treesitter can make the connection (and vice versa, if someone needs to look up upstream docs).

lewis6991 replied:
Good point actually, because TS already has the concept of "parser". God dammit.

Are we ok adding a `'lang'` bufoption in the presence of `'syntax'` and `'filetype'`?

clason replied:
Not ideal; I did prefer `parser` for it. I wish we had native table options already...

(Also, remember that it's n by m: multiple parsers may parse a filetype, and you might use different ones for different buffers -- for example upstream C and our forked C.)

But I don't really have a better suggestion at the moment.

lewis6991 replied:
Ok, I'll stick with `parser` for now. It's ok if the user mentally thinks of this as a TS parser.

lewis6991 commented on runtime/doc/treesitter.txt:
```diff
@@ -547,6 +547,21 @@ get_captures_at_pos({bufnr}, {row}, {col})
     Return: ~
         table[] List of captures `{ capture = "name", metadata = { ... } }`
 
+get_lang({opts})                                   *vim.treesitter.get_lang()*
```
Yeah, just got hit with this myself.

clason replied:
The comment parser is a scourge anyway...

lewis6991 replied:
Removed `get_lang` from this PR.

clason replied:
Since the code is now gone, can you remind us how to get the language string from a node? `get_node()` -> `tree()` -> parser -> `._lang`?

lewis6991 replied:
It's in the OP.

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -2203,6 +2203,27 @@ add({filetypes})                                          *vim.filetype.add()*
       • {filetypes}  (table) A table containing new filetype maps (see
                      example).
 
+get_option({filetype}, {option})                   *vim.filetype.get_option()*
```
Related vim todo for reference: https://github.com/vim/vim/blob/e764d1b4219e6615a04df1c3a6a5c0210a0a7dac/runtime/doc/todo.txt#L1036-L1038

```
Add options_default() / options_restore() to set several options to Vim
defaults for a plugin. Comments from Zyx, 2017 May 10.
Perhaps use a vimcontext / endvimcontext command block.
```

justinmk commented on runtime/lua/vim/filetype/options.lua:
```diff
@@ -0,0 +1,122 @@
+--- @param filetype string
+local function update_ft_option_cache(filetype)
+  local new_ftautos = #api.nvim_get_autocmds({ event = 'FileType' })
+  if new_ftautos ~= ft_autocmd_num then
```
Nice approach

seblj commented on runtime/lua/vim/filetype/options.lua:
```diff
@@ -0,0 +1,122 @@
+
+  local ret = vim.bo[buf][option] ---@type string|integer|boolean
+
+  api.nvim_buf_delete(buf, {})
```
This seems to cause a problem for me in some situations. I get an error on `E565: Not allowed to change text or change window`. I use vim-commentary and have an expr keymap that overrides the default one that just calculates the commentstring before returning `<Plug>Commentary`. It could maybe have something to do with me calling the function at a wrong time, but it should probably be documented if it's not allowed to get an option in some cases(?)

seblj replied:
comment.lua

```lua
local api = vim.api

local function noautocmd(f)
	local ei = vim.o.eventignore
	vim.o.eventignore = "all"
	f()
	vim.o.eventignore = ei
end

local function get_option(filetype, option)
	-- Some filetypes might not set 'commentstring' so we may end up
	-- returning 'commentstring' for another filetype if the buffer is re-used.
	local buf = api.nvim_create_buf(false, true)

	-- Prevent syntax/ from being loaded
	vim.b[buf].ts_highlight = true

	-- Change the filetype without triggering any autocmds
	noautocmd(function()
		vim.bo[buf].filetype = filetype
	end)

	-- Trigger the FileType autocmd in ftplugin.vim
	api.nvim_buf_call(buf, function()
		-- TODO(lewis6991): should we just execute filetypeplugin or all of FileType?
		-- Cache is only coherent for ftplugin but users might want to configure filetypes in their
		-- config.via basic FileType autocmds.
		-- api.nvim_exec_autocmds('FileType', { group = 'filetypeplugin', buffer = buf })
		api.nvim_exec_autocmds("FileType", { buffer = buf })
	end)

	local ret = vim.bo[buf][option] ---@type string|integer|boolean

	api.nvim_buf_delete(buf, {})

	return ret
end

local function calculate_commentstring()
	return get_option("lua", "commentstring")
end

vim.keymap.set("n", "gcc", function()
	vim.bo.commentstring = calculate_commentstring()
	return "<Plug>CommentaryLine"
end, { expr = true })
```

`nvim -u comment.lua test.lua`.

Insert some text and try `gcc`

lewis6991 replied:
Works for me. Are you sure you don't have any wacky FileType autocmds? Can you reproduce this in `nvim --clean`?

lewis6991 replied:
Nevermind, can repro.

lewis6991 replied:
How about if you uncomment this:

```lua
    -- api.nvim_exec_autocmds('FileType', { group = 'filetypeplugin', buffer = buf })
```

and remove the line below?

lewis6991 replied:
Also wrap the buf deletion in schedule:

```lua
  -- Need to schedule the buf deletion in case we are in an expr mapping
  vim.schedule(function()
    api.nvim_buf_delete(buf, {})
  end)
```

seblj replied:
Wrapping it in vim.schedule did the trick for me. Both of the exec_autocmds seem to work then

lewis6991 replied:
Also you can do this without an `expr` mapping.

```lua
vim.keymap.set("n", "gcc", function()
  local lang = get_lang() or vim.bo.filetype
  vim.bo.commentstring = get_ft_option(lang, 'commentstring')
  vim.cmd.normal(vim.api.nvim_replace_termcodes("<Plug>CommentaryLine", true, false, true))
end)
```

lewis6991 replied:
> Wrapping it in vim.schedule did the trick for me. Both of the exec_autocmds seem to work then

The problem is, we can't predict what will be in a `FileType` autocmd. For example, I had `vim.cmd.normal'gx'` which triggered a similar warning.

 So we will need to document this shouldn't be used in expr keymaps (though it might work).

seblj replied:
> > Wrapping it in vim.schedule did the trick for me. Both of the exec_autocmds seem to work then
> 
> The problem is, we can't predict what will be in a `FileType` autocmd. For example, I had `vim.cmd.normal'gx'` which triggered a similar warning.
> 
> So we will need to document this shouldn't be used in expr keymaps (though it might work).

Ah okay, will switch to the replace_termcodes variant👍 thanks!

lewis6991 replied:
@justinmk (and @bfredl) do you think it would be reasonable to override `textlock` and `ex_normal_lock` within `get_option()` because "we know what we're doing" (™) .

`eval_map_expr()` has this:

```c
  // Forbid changing text or using ":normal" to avoid most of the bad side
  // effects.  Also restore the cursor position.
  textlock++;
  ex_normal_lock++;
```

Which makes sense in the general case, but here could it be ok it bypass this? Especially since the mapping isn't really running in the scratch buffer.

Otherwise, we can just document the restriction.

justinmk replied:
> reasonable to override textlock and ex_normal_lock within get_option() because "we know what we're doing" (™) .

Yeah that's a good idea. We should do that regardless of where we end up with this PR and https://github.com/neovim/neovim/pull/22681

lewis6991 committed:
feat(api): add filetype option nvim_get_option_value

- Also adjust the expr-mapping behaviour so normal commands and text
  changes are allowed in internal dummy buffers.

zeertzjq commented on src/nvim/edit.c:
```diff
@@ -1245,7 +1245,9 @@ bool edit(int cmdchar, bool startln, long count)
   // caller of getcmdline() may get confused.
   // Don't allow recursive insert mode when busy with completion.
-  if (textlock != 0 || ins_compl_active() || compl_busy || pum_visible()) {
+  // Allow in dummy buffers since they are only used internally
+  if (textlock != 0 || ins_compl_active() || compl_busy || pum_visible()
+      || (expr_map_lock == 0 && !(curbuf->b_flags & BF_DUMMY))) {
```
What is the relation of this change to what this PR does?

mg979 commented:
Maybe it's a bit late, and at the risk of sounding overzealous, but is it safe to trigger FileType even on a dummy buffer? It could try to attach Lsp clients, or start servers, run external processes etc.

lewis6991 commented:
Maybe, I'm not sure. What's the first that could happen? However if we can't run the `FileType` autocmd safely, then there's no way we can safely use `ftplugin` as a namespace for filetype options.

mg979 commented:
An alternative approach to get filetype options could be to have lua ftplugins return a table, instead of nothing, but it's a completely different thing, and it would need all runtime ftplugins to be rewritten in lua, or at least add lua ftplugins for each of them, and have them return tables with options or anything that you'd want cached. Eventually I think rewriting them completely in lua wouldn't be a bad thing.

clason commented:
> An alternative approach to get filetype options could be to have lua ftplugins return a table, instead of nothing, but it's a completely different thing, and it would need all runtime ftplugins to be rewritten in lua, or at least add lua ftplugins for each of them, and have them return tables with options or anything that you'd want cached. Eventually I think rewriting them completely in lua wouldn't be a bad thing.

Completely out of scope. That would imply a hard fork of almost all of our runtime, which is a non-goal for Neovim.

mg979 commented:
> or at least add lua ftplugins for each of them, and have them return tables with options or anything that you'd want cached

This wouldn't require modifying the original scripts, but it's a lot of new files. Anyway, just saying what could be a different approach.

clason commented:
> This wouldn't require modifying the original scripts, but it's a lot of new files.

It's a lot of files that need to be manually kept in sync with upstream's runtime updates. Just saying why this is not an approach we'll even consider. 

(In fact, `filetype.lua` actually already supports returning filetype options (and does so for some), because this is one of the -- many -- ways Vim sets these options.)

clason commented:
I think this design is correct and useful; I don't see how the potential of side effects can be reduced without making the functionality harder to reason about. All it needs is a clear documentation about what happens, including a clear warning about the possible side effects.

If needed, we could mark this option as "experimental", allowing us to tweak behavior over the 0.10 cycle (or, indeed, bump to 0.10 altogether). But I'd like to get this in and get feedback from actual use. ("Fear is the mind killer" and whatnot.)

lewis6991 commented:
I agree.

I want to add a couple more tests for error scenarios before merging, but if I drag my feet then feel free to merge.

clason commented:
Tests are always good; planned feature freeze is end of March, so if you have a bit of time for that over the next week, that would be great; otherwise nightly users get to play test (fun!)

lewis6991 commented:
My main priority is this and the `'lang'` option (which uses this) so we can remove/tweak API that was added this dev cycle before the release.

clason commented on runtime/doc/api.txt:
```diff
@@ -1969,6 +1969,10 @@ nvim_get_option_value({name}, {*opts})               *nvim_get_option_value()*
                   Implies {scope} is "local".
+                • filetype: |filetype|. Used to get the default option for a
+                  specific filetype. Cannot be used with any other option.
+                  Note: this is expensive, it is recommended to cache this
```
I think we should be a bit more explicit here, e.g.,

"Note: This will trigger all filetype plugins and FileType autocommands for the corresponding filetype."

If more explanation is needed, it's probably better to add a short paragraph on the option to the main docstring body.

clason commented on runtime/doc/lua.txt:
```diff
@@ -2206,6 +2206,27 @@ add({filetypes})                                          *vim.filetype.add()*
+      vim.filetype.get_option('vim', 'commentstring')
+<
+
+    Note this function maintains an internal cache and is not guaranteed to
+    fully respect dynamic FileType autocmds and ftplugin files.
```
Maybe add a callout to `nvim_get_option_value`? E.g.,

"Note: This uses |nvim_get_option_value()| but caches the result. This means ftplugins and FileType autocommands are only triggered once and may not reflect later changes."

lewis6991 commented:
The issue needs to be analyzed and understood. The dummy buffer is purely an implementation detail, not something you should be messing with.

clason commented:
You can replicate everything nvim-lspconfig does with `vim.lsp.start` in a `FileType` autocommand.

mhanberg commented:
I will also try to come up with a repro.

I was attempting to replicate the commentstring example from the description and when i have codelens auto refreshing, i get tons of errors about an invalid buffer id.

it's also extremely slow (each time you move the cursor takes like 2 seconds), but that is merely tangential.

Thanks for all the hard work on Neovim, y'all have been on fire recently!!

clason commented:
yeah, the snippet is not the best way to realize that (you'd query the commentstring on demand, not run a `CursorMoved` autocommand just in case) -- it's just for the sake of illustration

lewis6991 commented:
The whole point of `vim.filetype.get_option()` is that it's cached, so something in your system is invalidating it, or it's not working.

seblj commented:
This is the most minimal I was able to get it:

```lua
local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({
    "git",
    "clone",
    "--filter=blob:none",
    "https://github.com/folke/lazy.nvim.git",
    "--branch=stable", -- latest stable release
    lazypath,
  })
end
vim.opt.rtp:prepend(lazypath)

require("lazy").setup({
    {
        "nvim-treesitter/nvim-treesitter",
        config = function()
            require("nvim-treesitter.configs").setup({
                highlight = {
                    enable = true,
                },
            })
        end,
        priority = 100,
        build = ":TSUpdate",
    },

    -- LSP
    {
        "neovim/nvim-lspconfig",
        config = function()
            require("lspconfig")["lua_ls"].setup({})
        end,
    },
})

vim.keymap.set("n", "gcc", function()
    vim.filetype.get_option(vim.bo.ft, "commentstring")
end)
```

I created a new config with `NVIM_APPNAME`, and then used this as `init.lua`. Open up and let lazy install everything.

Then do: `nvim test.lua` and run `gcc`. You should see a message with `[LSP] Client with id 1 not attached to buffer 2`.

Notice that I use `priority = 100` on `nvim-treesitter` (default is 50). If I however set it to `priority = 100` on `nvim-lspconfig` instead, I don't get notified. I have no idea why the order of loading matters, but it looks like it does. I tried to make it even more minimal with adding to `rtp` myself with the order as above, but I couldn't reproduce the error message

lewis6991 commented:
Repro both issues (LSP and bad cache), should be fixed in #22753

seblj commented:
> @seblj you should never need a package manager for a minimal repro. Best to use this template from gitsigns:
> 
> ```lua
> 
>          vim.o.packpath = '/tmp/nvim/site'
> 
> 
> 
>          local plugins = {
> 
>            gitsigns = 'https://github.com/lewis6991/gitsigns.nvim',
> 
>            -- ADD OTHER PLUGINS _NECESSARY_ TO REPRODUCE THE ISSUE
> 
>          }
> 
> 
> 
>          for name, url in pairs(plugins) do
> 
>            local install_path = '/tmp/nvim/site/pack/test/start/'..name
> 
>            if vim.fn.isdirectory(install_path) == 0 then
> 
>              vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
> 
>            end
> 
>          end
> 
> 
> 
>          require('gitsigns').setup{
> 
>            debug_mode = true, -- You must add this to enable debug messages
> 
>            -- ADD GITSIGNS CONFIG THAT IS _NECESSARY_ FOR REPRODUCING THE ISSUE
> 
>          }
> 
> ```

I know... I didn't want to use it but couldn't get it to work without it. Thanks for the snippet from gitsigns! Will use it in the future
